CMake: Use modern CMake techniques to set toolchain options

hugueskamba opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Use `target_compile_options()`, `target_compile_definitions()`,
and `target_link_options()` to set toolchain options.
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [√] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [√] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

hugueskamba committed:
CMake: Use modern CMake techniques to set toolchain options

Use target_compile_options(), target_compile_definitions(),
and target_link_options() to set toolchain options.

0xc0170 reviewed (commented):
nice, this is what I will use as a base to fix the build type.

will review in detail on Monday

rwalton-arm commented on cmake/app.cmake:
```diff
@@ -1,23 +1,26 @@
 # SPDX-License-Identifier: Apache-2.0
 
+# Create Mbed OS library
+add_library(mbed-os OBJECT)
```
I'm not sure about this. We should be adding the targets in the `CMakeLists.txt` entry points. A script named `app.cmake` shouldn't be responsible for adding the `mbed-os` library target. I believe we should also be adding the `app` executable in the application `CMakeLists.txt`. This is the standard CMake pattern and I think we should try to stick to it.

0xc0170 replied:
this is because of adding linking options to a target down in Mbed OS . you can add compile options for a library but not linking options (they are for executable). @hugueskamba is that correct? 

As stated above, it might be better to just rework the log to provide series of commits with  reasoning to answer our questions.

hugueskamba replied:
You cannot add options (link or compile) to a target that has not yet been created hence why the executable and library creation statements were moved before the include statements (which call the `target_` functions down the line).

rwalton-arm commented on cmake/app.cmake:
```diff
@@ -1,23 +1,26 @@
+# Create application executable
+add_executable(app)
+
 include(.mbedbuild/mbed_config.cmake)
```
Would it be worth moving these includes down into the `mbed-os` entry point? I think we should aim to eliminate `app.cmake` as it's currently coupled to both the library and the application. We should try and break this coupling so the application can just `target_link_libraries(mbed-os)` and everything works. I believe the added indirection of `app.cmake`, which seems to set properties that should belong in the `mbed-os` library, will be confusing to developers who have used CMake before.

0xc0170 replied:
> Would it be worth moving these includes down into the mbed-os entry point? 

Can you elaborate ? 

app.cmake should look like after we do clean-up and fixes to linking :

```
include(.mbedbuild/mbed_config.cmake)
include(${MBED_ROOT}/cmake/toolchain.cmake)
include(${MBED_ROOT}/cmake/core.cmake)
include(${MBED_ROOT}/cmake/profile.cmake)
include(${MBED_ROOT}/cmake/env.cmake)
include(${MBED_ROOT}/cmake/util.cmake)

# Include Mbed OS main cmake
add_subdirectory(mbed-os)
```

This rest of the app.cmake to move to main Mbed OS cmake, no need to have app.cmake after all ? We will get there one step at the time. I would like to address the linking fixes soon to remove all that glue from app.cmake (if that is possible as some things are needed from  tools and some from Mbed OS)

rwalton-arm replied:
> This to move to main Mbed OS cmake so no need to have app.cmake ?

Yes, I'm wondering if the `include` statements can be moved to `mbed-os/CMakeLists.txt` and in the top level application `CMakeLists.txt` we can just do something like:

```
add_executable(app)
add_subdirectory(mbed-os)
target_link_libraries(app mbed-os)
```

If we have a plan to do this in an incremental way, that's great. I'm just thinking about what the end result should look like.

hugueskamba replied:
I will update the commit history so the logical changes are more transparent.

From the conversation above, I take we are not yet moving executable creation to the app CMake source file. Therefore the PR will stay as is. Correct?

rwalton-arm replied:
I'm fine with changing it in another PR.

ladislas replied:
agreed with @rwalton-arm about removing app.cmake

rwalton-arm reviewed (dismissed):
Looks pretty good. I have a couple of comments though. Also note it would've been easier to review if each logical change was in its own commit.

hugueskamba committed:
CMake: Ensure targets creation before usage in `target_` functions

Move target creation statements before include statements. The targets
must be created before they can be used in `target_` functions.

hugueskamba committed:
CMake: Issue a fatal error if a lower version of CMake is used

urutva commented on cmake/cores/Cortex-A9.cmake:
```diff
@@ -4,55 +4,48 @@
+            ${common_toolchain_options}
+    )
+
+    target_link_options(app
```
Core cmake scripts should avoid influencing link options for the app directly. Instead, it should set link options to `mbed-os` target with PUBLIC keyword so that cmake can propagate the link option to the consumer (app).
Same comment for other cores script changes as well.

hugueskamba replied:
Working on it. I initially thought that the command could only be used for shared libraries because it did not work. I must have something else wrong at the time. I will push something to update the first commit in the branch.

hugueskamba replied:
[Done](https://github.com/ARMmbed/mbed-os/compare/69b9a15a7753f73fe4744211d1bb895ebadc39a1..82358905db82d7c7d76c4cb49ccfe2a1eba87a95).
